### PR TITLE
Allow Dict type in c10 operators

### PR DESCRIPTION
**Pull request by smessmer**
Stack:
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #20383 [wip] torch::jit::RegisterOperators forwards to c10::RegisterOperators&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/D15300937/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #20514 Options based registration API&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/D15346348/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #20513 Infer schema for experimental ops&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/D15346349/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #20379 Allow nested lists/dicts in legacy operator API&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/D15287693/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #20374 Extend testAvailableArgTypes&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/D15298233/)
&nbsp;&nbsp;&nbsp;&nbsp;:black_circle:&nbsp; **#20373 Allow Dict type in c10 operators**&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/D15298235/)

- Add support for Dict<Key, Value> arguments and returns to c10 operators
- Add support for std::unordered_map<Key, Value> to the legacy API (but not to c10 kernels)

Differential Revision: [D15298235](https://our.internmc.facebook.com/intern/diff/D15298235/)

### CODE REVIEW

File: aten/src/ATen/core/op_registration/kernel_functor.h [hunk 1]
```diff
@@ -106,66 +123,111 @@ namespace detail {
...
     static_assert(guts::false_t<T>::value, "You tried to register a kernel with an unsupported return type: const char*. Please use std::string instead.");
   };
   template<class T>
-  struct return_type_to_ivalue_<T, guts::enable_if_t<std::is_integral<T>::value && !guts::typelist::contains<supported_primitive_arg_types, T>::value>> {
+  struct return_type_to_ivalue<T, guts::enable_if_t<std::is_integral<T>::value && !guts::typelist::contains<supported_primitive_arg_types, T>::value>> {
     static_assert(guts::false_t<T>::value, "You tried to register a kernel with an unsupported integral return argument type. Please use int64_t instead.");
   };
-
+  // legacy_return_type_to_ivalue is like return_type_to_ivalue but additionally
+  // allows a few deprecated types like std::vector.
```
Comment (li-roy) [inline aten/src/ATen/core/op_registration/kernel_functor.h]: Did you mean to have vector here as well?

File: aten/src/ATen/core/op_registration/kernel_functor.h [hunk 2]
```diff
@@ -82,6 +92,13 @@ namespace detail {
     // Users should use ArrayRef instead.
     static_assert(guts::false_t<std::vector<T>>::value, "You tried to register a kernel with an unsupported argument type: std::vector<T>. Please use c10::ArrayRef<T> instead.");
   };
+  template<class Key, class Value>
+  struct ivalue_to_arg_type<std::unordered_map<Key, Value>> {
+    // We don't support std::vector because that would prevent us from doing
```
Comment (li-roy) [inline aten/src/ATen/core/op_registration/kernel_functor.h]: Update this comment?

**Review by li-roy — APPROVED** [review]
approved with minor comments

File: aten/src/ATen/core/op_registration/kernel_functor.h (hunk 1 above)
Comment (smessmer) [reply aten/src/ATen/core/op_registration/kernel_functor.h]: ah no, vector is actually still allowed, possibly to be deprecated in a future diff. The comment should say "unordered_map". Fixed it.

**Pull request closed without merging**